### PR TITLE
Apply i18n for percentage values

### LINKED ISSUE #633: Percentage values are not localisable
Percentage values are expressed in different formats for different languages. For instance, some European languages like French does it in the format of `100 %`, and some languages like Turkish display the sign before the number, like `%100`. However, PackageKit components currently do not allow l10n of percentage values, creating an incoherent look among the localised strings. It would be nice to have them exposed for localisation.

### PR DESCRIPTION
**Pull request by bitigchi**
Some languages like Turkish and French (and more) uses another percentage format other than 100% (%100 and 100 % respectively). Lack of i18n on UI strings causes an incohesive look, and wrong grammar-wise.

Fixes #633.

### CODE REVIEW

**Review by ximion — CHANGES_REQUESTED** [review]

File: client/pk-offline-update.c [hunk 1]
```diff
@@ -154,11 +154,13 @@ pk_offline_update_progress_cb (PkProgress *progress,
 		if (role == PK_ROLE_ENUM_UPGRADE_SYSTEM) {
 			/* TRANSLATORS: this is the message we send plymouth to
 			 * advise of the new percentage completion when installing system upgrades */
-			msg = g_strdup_printf ("%s - %i%%", _("Installing System Upgrade"), percentage);
+			msg = g_strdup_printf (_("%s - %i%%"),
```
[the remark this answers is not in the document]
Comment (bitigchi) [reply client/pk-offline-update.c:157]: You're right, I should have been more thorough. Will modify accordingly.

File: lib/packagekit-glib2/pk-progress-bar.c [hunk 2]
```diff
@@ -132,7 +133,7 @@ pk_progress_bar_draw (PkProgressBar *self, gint percentage)
 		g_string_append (str, " ");
 	g_string_append (str, "] ");
 	if (percentage >= 0 && percentage < 100)
-		g_string_append_printf (str, "(%i%%)  ", percentage);
+		g_string_append_printf (str, _("(%i%%)  "), percentage);
```
Comment (bitigchi) [inline lib/packagekit-glib2/pk-progress-bar.c:136]: Are trailing spaces really needed here?  AFAICS, it's already at the end of the progress line.

Comment (bitigchi) [inline lib/packagekit-glib2/pk-progress-bar.c:136]: Are trailing spaces really needed here?  AFAICS, it's already at the end of the progress line.

**Comment by bitigchi** [comment]
Ping. I would appreciate feedback on the updates.

**Comment by bitigchi** [comment]
May I ping on this one?

File: client/pk-offline-update.c [hunk 3]
```diff
@@ -150,15 +150,19 @@ pk_offline_update_progress_cb (PkProgress *progress,
 			return;
 		sd_journal_print (LOG_INFO, "percentage %i%%", percentage);
 
+		g_autofree gchar *tmp_perc = NULL;
+		/* TRANSLATORS: this is a percentage value we use in messages, e.g. "90%" */
+		tmp_perc = g_strdup_printf (_("%i%%"), percentage);
```
Comment (sidt4) [inline client/pk-offline-update.c:155]: Minor nitpick.

`tmp_perc` -> `percent_str` ( inline with `role_str`, `action_str` etc )

File: lib/packagekit-glib2/pk-progress-bar.c [hunk 4]
```diff
@@ -131,10 +133,13 @@ pk_progress_bar_draw (PkProgressBar *self, gint percentage)
 	for (i = 0; i < self->priv->size - section; i++)
 		g_string_append (str, " ");
 	g_string_append (str, "] ");
-	if (percentage >= 0 && percentage < 100)
-		g_string_append_printf (str, "(%i%%)  ", percentage);
-	else
+	if (percentage >= 0 && percentage < 100) {
+		/* TRANSLATORS: this is a percentage value we use in messages, e.g. "90%. */
+		g_string_append_printf (str, _("(%i%%)"), percentage);
```
Comment (sidt4) [inline lib/packagekit-glib2/pk-progress-bar.c:138]: do we need `%i` (signed) format specifier ?

File: lib/packagekit-glib2/pk-progress-bar.c [hunk 5]
```diff
@@ -172,10 +177,13 @@ pk_progress_bar_pulse_bar (PkProgressBar *self)
...
 	g_string_append (str, "] ");
-	if (self->priv->percentage >= 0 && self->priv->percentage != PK_PROGRESS_BAR_PERCENTAGE_INVALID)
-		g_string_append_printf (str, "(%i%%)  ", self->priv->percentage);
-	else
+	if (self->priv->percentage >= 0 && self->priv->percentage != PK_PROGRESS_BAR_PERCENTAGE_INVALID) {
+		/* TRANSLATORS: this is a percentage value we use in messages, e.g. "90%. */
+		g_string_append_printf (str, _("(%i%%)"), self->priv->percentage);
+		g_string_append_printf (str, "  ");
+	} else {
 		g_string_append (str, "        ");
```
Comment (sidt4) [inline lib/packagekit-glib2/pk-progress-bar.c:185]: From the MR description it appears that there could be extra spaces in the percentage string depending upon the locale. So, this needs to be done dynamically, I guess.

File: lib/packagekit-glib2/pk-progress-bar.c [hunk 6]
```diff
@@ -131,10 +133,13 @@ pk_progress_bar_draw (PkProgressBar *self, gint percentage)
 	for (i = 0; i < self->priv->size - section; i++)
 		g_string_append (str, " ");
 	g_string_append (str, "] ");
-	if (percentage >= 0 && percentage < 100)
-		g_string_append_printf (str, "(%i%%)  ", percentage);
-	else
+	if (percentage >= 0 && percentage < 100) {
+		/* TRANSLATORS: this is a percentage value we use in messages, e.g. "90%. */
```
Comment (sidt4) [inline lib/packagekit-glib2/pk-progress-bar.c:137]: "90%.   -> "90%"

File: lib/packagekit-glib2/pk-progress-bar.c [hunk 7]
```diff
@@ -172,10 +177,13 @@ pk_progress_bar_pulse_bar (PkProgressBar *self)
 	for (i = 0; i < (gint) (self->priv->size - self->priv->pulse_state.position - 1); i++)
 		g_string_append (str, " ");
 	g_string_append (str, "] ");
-	if (self->priv->percentage >= 0 && self->priv->percentage != PK_PROGRESS_BAR_PERCENTAGE_INVALID)
-		g_string_append_printf (str, "(%i%%)  ", self->priv->percentage);
-	else
+	if (self->priv->percentage >= 0 && self->priv->percentage != PK_PROGRESS_BAR_PERCENTAGE_INVALID) {
+		/* TRANSLATORS: this is a percentage value we use in messages, e.g. "90%. */
```
Comment (sidt4) [inline lib/packagekit-glib2/pk-progress-bar.c:181]: "90%.   -> "90%"

File: client/pk-offline-update.c [hunk 8]
```diff
@@ -150,15 +150,19 @@ pk_offline_update_progress_cb (PkProgress *progress,
 			return;
 		sd_journal_print (LOG_INFO, "percentage %i%%", percentage);
 
+		g_autofree gchar *tmp_perc = NULL;
```
Comment (sidt4) [inline client/pk-offline-update.c:153]: This declaration needs to be moved to the top, else you'll get `mixed declarations and code` warning from the compiler.

File: client/pk-offline-update.c (hunk 3 above)
Comment (bitigchi) [reply client/pk-offline-update.c]: Done.

File: lib/packagekit-glib2/pk-progress-bar.c (hunk 4 above)
Comment (bitigchi) [reply lib/packagekit-glib2/pk-progress-bar.c]: Replaced with `%d`.

File: lib/packagekit-glib2/pk-progress-bar.c (hunk 5 above)
Comment (bitigchi) [reply lib/packagekit-glib2/pk-progress-bar.c:185]: I think the issue stems from the application of `pk_info_enum_to_localised_present` and `pk_info_enum_to_localised_past`. I need to look a bit deeper when I have more time. This particular line should not contribute to the misalignment if I read the code correctly.

File: lib/packagekit-glib2/pk-progress-bar.c (hunk 6 above)
Comment (bitigchi) [reply lib/packagekit-glib2/pk-progress-bar.c]: Done

File: lib/packagekit-glib2/pk-progress-bar.c (hunk 7 above)
Comment (bitigchi) [reply lib/packagekit-glib2/pk-progress-bar.c]: Done.

File: client/pk-offline-update.c (hunk 8 above)
Comment (bitigchi) [reply client/pk-offline-update.c]: Done.

File: lib/packagekit-glib2/pk-progress-bar.c (hunk 4 above)
Comment (ximion) [reply lib/packagekit-glib2/pk-progress-bar.c]: Either would have been fine here, when used in printf() there's no real difference between them anymore.

File: lib/packagekit-glib2/pk-progress-bar.c (hunk 5 above)
Comment (ximion) [reply lib/packagekit-glib2/pk-progress-bar.c:185]: Is there any chance that translators will insert a lot of spaces in the translated string or make this really long? TBH, the space alignment on stdout is pretty bad already, so it probably wouldn't make much of a difference, but we should at least not make things worse ^^
If translators won't make the % translation long and e.g. just change the ordering, then this patch is fine as-is.

Comment (bitigchi) [reply lib/packagekit-glib2/pk-progress-bar.c:185]: The longest variant is `100 %`, which is used in several European countries. I think it should be alright.

**Comment by bitigchi** [comment]
@ximion would it be possible to get a final word on this? I'd love to make this ready for a possible release before Plasma 6. Otherwise it would probably take a long time to ship this on distros.

**Review by ximion — APPROVED** [review]
Sorry for the long delay, and thank you for making the requested changes! This looks good now, thank you for the patch!

**Comment by piotrdrag** [comment]
You need to add `lib/packagekit-glib2/pk-progress-bar.c` to `po/POTFILES.in` for this to take effect.

**Comment by bitigchi** [comment]
@piotrdrag, thanks for the reminder. Fixed it in #725.

**Pull request merged**